tests/test_publisher.py

```diff
@@ -0,0 +1,98 @@
+import os
+import queue
+import threading
+from itertools import chain
+
+import pytest
+
+from lektor.publisher import _join_pipes_using_select
+from lektor.publisher import _join_pipes_using_threads
+from lektor.publisher import Command
+
+
+def test_Command_triggers_no_warnings(recwarn):
+    # This excercises the issue where publishing via rsync resulted
+    # in ResourceWarnings about unclosed streams.
+
+    # This is essentially how RsyncPublisher runs rsync.
+    with Command(["echo"], env={}) as client:
+        for _ in client:
+            pass
+    del client
+
+    for warning in recwarn.list:
+        print(warning)
+    assert len(recwarn) == 0
+
+
+def iter_in_thread(it):
+    q = queue.Queue()
+    eof = object()
+
+    def run():
+        for item in chain(it, [eof]):
+            q.put(item)
+
+    threading.Thread(target=run, daemon=True).start()
+    while True:
+        try:
+            item = q.get(timeout=0.1)
+        except queue.Empty:
+            item = "[TIMED OUT]"
+        else:
+            if item is eof:
+                break
+        yield item
+
+
+class Test__join_pipes:
+    # pylint: disable=no-self-use
+
+    @pytest.fixture(params=[_join_pipes_using_threads, _join_pipes_using_select])
+    def join_pipes(self, request):
+        if request.param is _join_pipes_using_select and os.name == "nt":
+            pytest.skip("_join_pipes_using_select does not work on Windows")
+        return request.param
+
+    @pytest.fixture
+    def pipes(self):
+        class Streams(tuple):
+            def close(self):
+                for stream in self:
+                    if not stream.closed:
+                        stream.close()
+
+        pipes = [os.pipe() for n in range(2)]
+        readers = Streams(open(r, "rb") for r, w in pipes)
+        writers = Streams(open(w, "wb", buffering=0) for r, w in pipes)
+        try:
+            yield readers, writers
+        finally:
+            writers.close()
+            readers.close()
+
+    def test(self, join_pipes, pipes):
```

Comment:
In general I prefer the tests to be flat. I think these could be `test_join_pipes_ordering` (if that is what it tests) and `test_join_pipes_prevents_partial_line_deadlock`. I know `pytest` supports both methods but I consider not grouping by class to be a bit more standard.